### ISSUE TITLE
Memory Leak Issue in YOLO11-JDE Inference on 4K Videos

### ISSUE THREAD

**Issue by jayer95** [inakierregueab/YOLO11-JDE#12]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Dear inakierregueab,

I have noticed that during inference with your YOLO11-JDE, the memory usage continuously increases until it eventually fills up and crashes. I am using track.py for video inference, and I am using the model you provided: YOLO11s_JDE-CHMOT17-64b-100e_TBHS_m075_1280px.pt.

For inference, I am using a 4K video at 30 FPS for 5 minutes. Around the 5000th frame, my system’s non-GPU memory (128GB) gets completely filled and crashes. Does this issue occur on your side as well?

If necessary, I can provide my test video (4K/30FPS/5 minutes)

Thank you.

### Environment

Python 3.12.7
CUDA 12.6
NVIDIA 550.120

Package                  Version
------------------------ ------------
absl-py                  2.1.0
autocommand              2.2.2
backports.tarfile        1.2.0
blinker                  1.9.0
certifi                  2024.8.30
charset-normalizer       3.4.0
click                    8.1.8
coloredlogs              15.0.1
contourpy                1.3.0
cycler                   0.12.1
einops                   0.8.1
faiss-cpu                1.10.0
filelock                 3.13.1
Flask                    3.1.0
flatbuffers              24.3.25
fonttools                4.54.1
fsspec                   2024.2.0
grpcio                   1.70.0
humanfriendly            10.0
idna                     3.10
importlib_metadata       8.0.0
importlib_resources      6.4.0
inflect                  7.3.1
itsdangerous             2.2.0
jaraco.collections       5.1.0
jaraco.context           5.3.0
jaraco.functools         4.0.1
jaraco.text              3.12.1
Jinja2                   3.1.3
joblib                   1.4.2
kiwisolver               1.4.7
lapx                     0.5.11.post1
Markdown                 3.7
MarkupSafe               2.1.5
matplotlib               3.9.2
more-itertools           10.3.0
mpmath                   1.3.0
networkx                 3.2.1
numpy                    1.26.3
nvidia-cublas-cu12       12.4.5.8
nvidia-cuda-cupti-cu12   12.4.127
nvidia-cuda-nvrtc-cu12   12.4.127
nvidia-cuda-runtime-cu12 12.4.127
nvidia-cudnn-cu12        9.1.0.70
nvidia-cufft-cu12        11.2.1.3
nvidia-curand-cu12       10.3.5.147
nvidia-cusolver-cu12     11.6.1.9
nvidia-cusparse-cu12     12.3.1.170
nvidia-nccl-cu12         2.21.5
nvidia-nvjitlink-cu12    12.4.127
nvidia-nvtx-cu12         12.4.127
onnx                     1.17.0
onnxruntime              1.20.1
onnxruntime-gpu          1.20.1
onnxslim                 0.1.41
opencv-python            4.10.0.84
packaging                24.1
pandas                   2.2.3
pillow                   10.2.0
pip                      24.2
platformdirs             4.2.2
protobuf                 5.28.3
psutil                   6.1.0
py-cpuinfo               9.0.0
pyparsing                3.2.0
python-dateutil          2.9.0.post0
pytorch-metric-learning  2.8.1
pytz                     2024.2
PyYAML                   6.0.2
requests                 2.32.3
scikit-learn             1.6.1
scipy                    1.14.1
seaborn                  0.13.2
setuptools               75.1.0
six                      1.16.0
sympy                    1.13.1
tabulate                 0.9.0
tensorboard              2.19.0
tensorboard-data-server  0.7.2
termcolor                2.5.0
threadpoolctl            3.5.0
tomli                    2.0.1
torch                    2.5.0+cu124
torchaudio               2.5.0+cu124
torchvision              0.20.0+cu124
tqdm                     4.66.5
triton                   3.1.0
typeguard                4.3.0
typing_extensions        4.9.0
tzdata                   2024.2
ultralytics              8.3.75
ultralytics-thop         2.0.9
urllib3                  2.2.3
Werkzeug                 3.1.3
wheel                    0.44.0
yacs                     0.1.8
zipp                     3.19.2

### Minimal Reproducible Example

from collections import defaultdict

import cv2
import numpy as np
from tqdm import tqdm

from ultralytics import YOLO

# Load the model and run the tracker with a custom configuration file
model = YOLO("YOLO11s_JDE-CHMOT17-64b-100e_TBHS_m075_1280px.pt", task="jde")
results = model.track(
    source="taipei4k_5min_30fps.mp4",
    tracker="jdetracker.yaml",
    show=True,  # show the video while tracking
    persist=True,   # if input is frames, persist=True will track objects across frames
    # Plus any other YOLO arguments for prediction
    # imgsz=(608,1088),
    save=True,
)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by inakierregueab** [comment]
Hi there! Sorry for the late response. The tracking framework is the one from Ultralytics repo. I have just modify it to work with my custom tracker algorithm and include the Re-ID embeddings in the loop. Have you experienced this issues with their official repo code? From my side I haven't encountered this issue after several tests. I recommend you to build your custom tracking pipeline if you have very specific needs, such as online tracking. In which you do not rely on the track method of the YOLO class. Instead you do the data loading and feed the tracker manually (like in YOLO11-JDE/tracker/evaluation/mot_callback.py), maybe even using multi-thread. In any case, I'd be happy to test with your video if you share it.

**Comment by jayer95** [comment]
@inakierregueab 

Thank you for your reply.
I was able to reproduce the test scenario on my end.
I cloned the YOLO11-JDE repository from GitHub and made almost no modifications.
The following video demonstrates the issue:
https://youtu.be/PTsd-jCzrnA?si=PL9wNfRG-mjwOVl5

Video URL:
https://drive.google.com/file/d/1d_6sttmZRfdTJ1zPCG5_4yOB7m1CqvMI/view?usp=sharing

**Comment by inakierregueab** [comment]
I'll try to reproduce the error but I do strongly recommend you to test persistent=False and also stream=True. In the past I have tested with long videos but always passing frames one by one, not feeding the video directly. In those cases, persistent could be set to False (check if the IDs are well maintained). Additionally when using stream=False (which is the default), all results are stored in memory, which can lead to higher memory usage for long videos, I think this may be the cause of your problem, since you are feeding the video directly and stream=False, all results are being stored and with a long sequence it will end up crashing. I suggest you to read the official Ultralytics documentation (https://docs.ultralytics.com/modes/track/).

I also want to point out that performing inference on 4k images without resizing may severely impacts performance (since the model was trained at 1280 pixels, the gap is huge). And most important, it increases memory consumption. If possible, consider resizing frames to a more manageable resolution before inference, especially if your application doesn’t require full 4K details.

Again, I strongly recommend you to read the video and pass frame by frame, like in YOLO11-JDE/tracker/evaluation/mot_callback.py, this is a standard practice that I have always followed when deploying YOLO models for tracking.

**Comment by inakierregueab** [comment]
Hi, the final model has been trained at 1280 and performs inference at 1280. You may experiment between different values based on the computation budged and the performance offered, especially if you are dealing with small objects. I'd say that the higher the size the better in terms of performance for both detection and Re-ID, but it might increase latency.

**Comment by jayer95** [comment]
@inakierregueab 
Thank you for providing the pre-trained model,
it's been really helpful for the POC I'm currently working on.
